File: src/retry.ts
// Bounded exponential backoff for transient upstream failures.
//
// Used inside UpstreamClient.send to give network blips and transient HTTP
// statuses (408/429/502/503/504) a few retries. Does NOT retry on 401 (the
// DPoP/nonce recovery in upstream.ts owns that) or any other 4xx, and does NOT
// retry a deliberate timeout (AbortError) — that deadline has already passed.
export interface BackoffOptions {
  retries?: number;
  baseMs?: number;
  /** Injectable delay (tests pass a no-op). */
  sleep?: (ms: number) => Promise<void>;
}

const TRANSIENT_STATUS = new Set([408, 429, 502, 503, 504]);

const defaultSleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Invoke `fn` (a fetch-like call returning a Response), retrying on transient
 * statuses and network errors with exponential backoff. Returns the last
 * Response once retries are exhausted; rethrows the last error on persistent
 * network failure.
 */
export async function withBackoff(
  fn: () => Promise<Response>,
  opts: BackoffOptions = {},
): Promise<Response> {
  const retries = opts.retries ?? 2;
  const baseMs = opts.baseMs ?? 200;
  const sleep = opts.sleep ?? defaultSleep;

  let lastError: unknown;
  for (let attempt = 0; attempt <= retries; attempt += 1) {
    try {
      const res = await fn();
      if (attempt < retries && TRANSIENT_STATUS.has(res.status)) {
        await sleep(baseMs * 2 ** attempt);
        continue;
      }
      return res;
    } catch (err) {
      lastError = err;
      const isAbort = err instanceof Error && err.name === "AbortError";
      if (isAbort || attempt >= retries) throw err;
      await sleep(baseMs * 2 ** attempt);
    }
  }
  throw lastError; // unreachable: the loop returns or throws
}

File: internal/upstream/retry.go
// Package upstream proxies MCP JSON-RPC to the adapter over HTTPS with DPoP,
// the Go port of src/upstream.ts and src/retry.ts.
package upstream

import (
	"context"
	"errors"
	"net/http"
	"time"
)

var transientStatus = map[int]bool{408: true, 429: true, 502: true, 503: true, 504: true}

// WithBackoff invokes fn (a fetch-like call), retrying transient statuses
// (408/429/502/503/504) and network errors with exponential backoff. It does
// NOT retry a context deadline/cancel (the deadline already passed) or any other
// 4xx. Returns the last response when retries are exhausted; returns the last
// error on persistent network failure.
func WithBackoff(fn func() (*http.Response, error), retries, baseMs int, sleep func(time.Duration)) (*http.Response, error) {
	if sleep == nil {
		sleep = time.Sleep
	}
	var lastErr error
	for attempt := 0; attempt <= retries; attempt++ {
		res, err := fn()
		if err != nil {
			lastErr = err
			if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) || attempt >= retries {
				return nil, err
			}
			sleep(backoff(baseMs, attempt))
			continue
		}
		if attempt < retries && transientStatus[res.StatusCode] {
			res.Body.Close()
			sleep(backoff(baseMs, attempt))
			continue
		}
		return res, nil
	}
	return nil, lastErr
}

func backoff(baseMs, attempt int) time.Duration {
	return time.Duration(baseMs*(1<<attempt)) * time.Millisecond
}
